pkg/net/rpc.go

```diff
@@ -106,222 +72,51 @@ func (o *Overlay) matchGossip(s string) bool {
-	}
-
-	return binary.Write(s, binary.BigEndian, w.nextStep(ctx))
-}
-
-func (w randWalk) setWriteDeadline(ctx context.Context, s network.Stream) error {
-	if t, ok := ctx.Deadline(); ok {
-		return s.SetWriteDeadline(t)
-	}
-
-	return nil
-}
-
-func (w randWalk) nextStep(ctx context.Context) *sampleStep {
-	s := sampleStep{Depth: w.depth - 1}
-	if t, ok := ctx.Deadline(); ok {
-		s.Timeout = time.Until(t)
-	}
-	return &s
-}
-
-func (w randWalk) awaitResult(ctx context.Context, s network.Stream) error {
-	if err := w.setReadDeadline(ctx, s); err != nil {
-		return err
-	}
-
-	return w.readEnvelope(ctx, s)
-}
-
-func (w randWalk) readEnvelope(ctx context.Context, s network.Stream) error {
-	b, err := w.readPayload(ctx, s)
-	if err != nil {
-		return err
-	}
-
-	return w.deliver(ctx, b)
-}
-
-func (w randWalk) readPayload(ctx context.Context, s network.Stream) ([]byte, error) {
-	if err := w.setReadDeadline(ctx, s); err != nil {
-		return nil, err
-	}
-
-	return ioutil.ReadAll(s)
-}
-
-func (w randWalk) setReadDeadline(ctx context.Context, s network.Stream) error {
-	if t, ok := ctx.Deadline(); ok {
-		return s.SetReadDeadline(t)
-	}
-
-	return nil
-}
-
-func (w randWalk) deliver(ctx context.Context, b []byte) error {
-	_, rec, err := record.ConsumeEnvelope(b, peer.PeerRecordEnvelopeDomain)
-	if err != nil {
-		return err
-	}
-
-	return w.d.Deliver(ctx, rec.(*peer.PeerRecord))
-}
-
-type sampleStep struct {
-	Depth   uint8
-	Timeout time.Duration
-}
-
-func (s sampleStep) Deadline() time.Time {
-	if s.Timeout == 0 {
-		return time.Now().Add(defaultSampleStepTimeout)
-	}
-
-	return time.Now().Add(s.Timeout)
-}
-
-func (ss *sampleStep) RecvPayload(s network.Stream) error {
-	if err := s.SetReadDeadline(ss.Deadline()); err != nil {
-		return err
-	}
-
-	return binary.Read(s, binary.BigEndian, ss)
-}
-
-func (ss *sampleStep) Next(s network.Stream, h host.Host, peer peer.ID) stepper {
-	if ss.Depth > 1 {
-		return randWalk{
-			peer:  peer,
-			depth: ss.Depth - 1,
-			d:     deliveryStream{s: s, h: h},
-			sd:    h,
-			proto: s.Protocol(),
-		}
-	}
-
-	return randChoice{
-		peer: peer,
-		d:    deliveryStream{s: s, h: h},
-		h:    h,
-	}
-}
-
-type randChoice struct {
-	peer peer.ID
-	d    deliverer
-	h    host.Host
-}
-
-func (c randChoice) Step(ctx context.Context) error {
-	return c.d.Deliver(ctx, &peer.PeerRecord{
-		PeerID: c.peer,
-		Addrs:  c.h.Peerstore().Addrs(c.peer),
-	})
-}
-
 type gossip struct {
-	n     *neighborhood
-	sd    streamDialer
-	h     host.Host
-	peer  peer.ID
-	proto protocol.ID
+	n      *neighborhood
+	sd     streamDialer
+	h      host.Host
+	peerID peer.ID
+	proto  protocol.ID
 }
 
-func (g gossip) PushPull(ctx context.Context) ([]*peer.PeerRecord, error) {
-	s, err := g.sd.NewStream(ctx, g.peer, g.proto)
+func (g gossip) PushPull(ctx context.Context) (recordSlice, error) {
+	s, err := g.sd.NewStream(ctx, g.peerID, g.proto)
 	if err != nil {
 		return nil, err
 	}
 	defer s.Close()
+	gr, ctx := errgroup.WithContext(ctx)
```

Comment:
I don't know if the subcontext is necessary or desireable. 🤔 
If e.g. a push fails, there's no reason to cancel the pull, is there?

It seems like we could fail gracefully by allowing at least one of the two peers to successfully update their view.  Or does that somehow undermine the algorithm?

Reply:
In the article they said that `only push` lead to higher partitioning, while `only pull` lead to not allow new peers to join the network. This only happens if there are many `only pulls` or `only push`, which is highly unlikely. However, until we do some simulations I would leave it like this just in case.

Reply:
Ah ok, that makes a lot of sense.  100% agreed.

+1 for the belt-and-suspenders approach to safety.